### PR TITLE
Improve filter api javadoc

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Enhancement / new feature
- Documentation

### Description

* start package level javadoc for filters.  My intention is that we can use this to document higher level concerns then class and method level javadoc can reference the package level documentation.
* use @NonNull on the callable API surface.
* renamed some formal parameters for consistency.

### Additional Context

_Why are you making this pull request?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [x] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by k-wall** [comment]
My intent had been to use @NoneNull on the API surface where the user implements an interface e.g. ProduceRequestFilter.  My main aim of doing this was to get the IDE to warn the programmer that they were returning null.   However, this approach didn't work too well in Intellij. It generated warnings for every @Override that didn't carry the same @NoneNull annotation.  The made for a sea of yellow.


Screenshot 2023-08-11 at 14 49 46

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/package-info.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+/**
+ *
+ * <h2 id='implementing'>Implementing Filters</h2>
+ *
+ * <h3 id='implementing.filterResults'>Filter Results</h3>
+ * <p>Filter implementation must return a {@link java.util.concurrent.CompletionStage} containing a
+ * {@link io.kroxylicious.proxy.filter.FilterResult} object. It is the job of FilterResult to convey what
+ * message is to forwarded to the next filter in the chain (or broker /client if at the chain's beginning
```
Comment (robobario) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/package-info.java:14]: `forwarded` -> `be forwarded`

I feel like we should lead with the idea that the FilterResult is an instruction to the framework, as you said below. The FilterResult instructs the framework what to do after the user has finished doing their work in the protocol filter. Options including: forward a message, close connection, drop message. Though I get that we should emphasise the message forwarding part. So I'd rephrase it something like:

`It is the job of FilterResult to carry an instruction to the framework, describing what should be done when the custom protocol filter has finished operating on the message. The most common instruction is to forward the message to the next filter in the chain. Other options include closing the connection or dropping the message.`

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/package-info.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * or end).  It is also used to carry instructions such as indicating that the connection must be closed,
+ * or a message dropped.</p>
+ * <h3 id='implementing.deferringForward'>Deferring Forwards</h3>
+ * <p>If the filter returns a CompletionStage that is already completed normally, Kroxylicious will immediately
+ * perform the action described by the `FilterResult`.</p>
```
Comment (SamBarker) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/package-info.java:19]: slightly begs the question what we should do with stages which are immediately completed exceptionally...

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/package-info.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ * <p>The filter may return a `CompletionStage` that is not yet completed. When this happens, Kroxylicious will pause
+ * reading from the downstream (the Client writes will eventually block), and it begins to queue up in-flight
+ * requests/responses arriving at the filter.  This is done so that message order is maintained.  Once the
```
Comment (SamBarker) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/package-info.java:22]: Should we call out at this level that this is an expensive operation in terms of performance and should be used sparingly?

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/package-info.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+/**
+ *
+ * <h2 id='implementing'>Implementing Filters</h2>
+ *
```
Comment (SamBarker) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/package-info.java:10]: We should probably have some text here about:

- Thread safety
- The fact that filter instances are scoped to a specific client connection
- Maybe something about sharing state?

 Potentially could be left to an alternate PR.

**Review by SamBarker — APPROVED** [review]
LGTM

Thanks @k-wall

**Review by robobario — APPROVED** [review]
LGTM thanks @k-wall

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/package-info.java (hunk 4 above)
Comment (k-wall) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/package-info.java:10]: Raised as separate issue (#557)

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/package-info.java (hunk 3 above)
Comment (k-wall) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/package-info.java]: I've added the same paragraph that already appears in the Asciidoc.
